src/design_notes/general_coroutines.md

```diff
@@ -0,0 +1,258 @@
+  with.
+
+```rust
+// This is a coroutine
+|stuff| {
+  yield stuff + 1;
+  yield stuff + 2;
+  return stuff + 3;
+}
+
+// This is a generator
+stream! {
+  yield 1;
+  yield 2;
+  yield 3;
+}
+```
+
+## Coroutine trait
+
+- The coroutine syntax must produce implementations of some trait.
+- [RFC-2781][2] and [eRFC-2033][1] propose the `Generator` trait.
+- Note that Rust's coroutines and subroutines look the same from the outside:
+  take input, mutate state, produce output.
+- Thus, [MCP-49][3] proposes using the `Fn*` traits instead, including a new
+  `FnPin` for immovable coroutines.
+  - Hierarchy: `Fn` is `FnMut + Unpin` is `FnPin` is `FnOnce`.
+    - May not be *required* at the trait level (someone may someday find a use
+      to implementing `FnMut + !FnPin`) but all closures implement the traits in
+      this order.
+
+## Coroutine syntax
+
+- The closure syntax is reused for coroutines by [eRFC-2033][1], [RFC-2781][2],
+  and [MCP-49][3].
+- Commentators have suggested that the differences between coroutines and
+  closures under [eRFC-2033][1] and [RFC-2781][2] justify an entirely distinct
+  syntax to reduce confusion.
+- [MCP-49][3] fully reuses the *semantics* of closures, greatly simplifying the
+  design space and making the shared syntax obvious.
+
+## Taking input
+
+- The major disagreement between past proposals is whether to use "yield
+  expressions" or "magic mutation".
+  - Yield expression: `let after = yield output;`
+    - Used by [eRFC-2033][1]
+  - Magic mutation: `let before = arg; yield output; let after = arg;`
+    - Used by [RFC-2781][2] and [MCP-49][3]
+- Many people have a strong gut preference for yield expressions.
+  - In simple cases, Rust generally prefers to produce values as output from
+    expressions rather than by mutation of state ⇒ yield expressions *feel* more
+    Rusty.
+  - I think it might be a bit like the `.await` bikeshed? People didn't *feel*
+    like `.await` was correct but it turned out to be well-justified.
+  - Justification ⇒ in addition to reasons below, holding references to past
+    resume args is rare, often a logic error. Rust can use mutation checks to
+    catch and give feedback.
+- "Magic mutation" is a bit of a misnomer. The resume argument values are *not
+  being mutated.* The argument bindings are simply being reassigned across
+  yields.
+  - In a sense, argument bindings are reassigned in the exact same way across
+  returns.
+- "Yield expression" causes problems with first-resume input.
+  - [eRFC-2033][1] passes the first resume argument via a closure parameters
+    while later arguments are produced by `yield` expressions.
+  - This does not play nicely with alternate coroutine syntaxes like `coroutine
+    { }`.
+- "Magic mutation" passes arguments in the same way as any normal function call.
+  Although arguments can be moved into the coroutine witness when needed, they
+  do not bloat the state otherwise.
```

Comment:
Users of yield-expressions are not *required* to bind resume args at all. However `let x = yield` does seem to be the common pattern for yield-expressions (having arguments be named is really important for making coroutines at all legible) and in that case, if `x` has drop, it would get introduced into the witness.

By comparison, magic-mutation gives users named resume args *without* putting the args into the witness (unless the user explicitly re-binds the args for use after the following yield) since the bindings are overwritten with the latest args on every resume (old args having been dropped or moved during the last state).

I'll see if I can clarify this a bit more in the notes. It is probably the most important distinction to make but I find it hard to explain concisely. Lots of annotated examples maybe?